client/src/main/java/io/pravega/client/batch/impl/BatchClientImpl.java

```diff
@@ -67,53 +76,101 @@ private StreamInfo getStreamInfo(Stream stream) {
-        return listSegments(stream, new Date(0L));
+    public StreamSegmentsInfo getSegments(final Stream stream, final StreamCut fromStreamCut, final StreamCut toStreamCut) {
+        Preconditions.checkNotNull(stream, "stream");
+        return listSegments(stream, Optional.ofNullable(fromStreamCut), Optional.ofNullable(toStreamCut));
     }
 
-    private Iterator<SegmentInfo> listSegments(Stream stream, Date from) {
-        // modify iteration above but starting with a timestamp and ending with a break
-        Map<Segment, Long> segments = getAndHandleExceptions(controller.getSegmentsAtTime(new StreamImpl(stream.getScope(),
-                                                                                                         stream.getStreamName()),
-                                                                                          from.getTime()),
-                                                             RuntimeException::new);
-        SortedSet<Segment> result = new TreeSet<>();
-        result.addAll(segments.keySet());
-        StreamSegmentSuccessors successors = getAndHandleExceptions(controller.getSuccessors(new StreamCutImpl(stream, segments)),
-                RuntimeException::new);
+    @Override
+    public <T> SegmentIterator<T> readSegment(final SegmentRange segment, final Serializer<T> deserializer) {
+        return new SegmentIteratorImpl<>(inputStreamFactory, segment.asImpl().getSegment(), deserializer,
+                segment.asImpl().getStartOffset(), segment.asImpl().getEndOffset());
+    }
+
+    private StreamSegmentsInfo listSegments(final Stream stream, final Optional<StreamCut> startStreamCut,
+                                            final Optional<StreamCut> endStreamCut) {
+        //Validate that the stream cuts are for the requested stream.
+        startStreamCut.ifPresent(streamCut -> Preconditions.checkArgument(stream.equals(streamCut.asImpl().getStream())));
+        endStreamCut.ifPresent(streamCut -> Preconditions.checkArgument(stream.equals(streamCut.asImpl().getStream())));
+
+        // if startStreamCut is not provided use the streamCut at the start of the stream.
+        // if toStreamCut is not provided obtain a streamCut at the tail of the stream.
+        CompletableFuture<StreamCut> startSCFuture = startStreamCut.isPresent() ?
+                CompletableFuture.completedFuture(startStreamCut.get()) : fetchStreamCut(stream, new Date(0L));
+        CompletableFuture<StreamCut> endSCFuture = endStreamCut.isPresent() ?
+                CompletableFuture.completedFuture(endStreamCut.get()) : fetchTailStreamCut(stream);
+
+        //fetch the StreamSegmentsInfo based on start and end streamCuts.
+        CompletableFuture<StreamSegmentsInfo> streamSegmentInfo = startSCFuture.thenCombine(endSCFuture,
+                (startSC, endSC) -> getStreamSegmentInfo(startSC, endSC));
+        return getAndHandleExceptions(streamSegmentInfo, RuntimeException::new);
```

Comment:
What exceptions can we get here? What should the application expect?

Reply:
This method fetches the `startStreamCut` (current HEAD of the stream) and `endStreamCut`(Current TAIL of the stream) by using the controller client APIs(if not provided by the application). It also computes the segments between the two streamCuts using Controller APIs. These controller  APIs have their internal retry logic.

The exceptions observed /expected here are :
-- The exceptions observed by the controller client post the retry logic. 
-- Input Validation Errors: i.e. 
    a. Exceptions thrown by the controller due to an invalid stream/scope name.
    b. Exceptions thrown by the `getSegments` Controller API due to invalid or overlapping segments.